### ISSUE TITLE
connection to server is erroring

### ISSUE THREAD

**Issue by lewisl** [nim-lang/langserver#228]
The full error message is:
```
Client Nim Language Server: connection to server is erroring.
Cannot call write after a stream was destroyed
```

Then:
```
Request textDocument/signatureHelp failed.
```

Then:
```
Client Nim Language Server: connection to server is erroring.
write EPIPE
Shutting down server.
```

My environment:

MacBook Air M2
macOS Sonoma 14.6.1
Visual Studio Code version 1.92.2 (Universal)

The work around would appear to be either:
a) stop using the Nim team's maintained version of the nimlang plugin for VS Code and go back to Nimsaem's version
b) stop using the language server from either Nimlang plugin and go directly to nimsuggest.  All these recommendations say, "Oh, nimsuggest crashes; must use lsp."  The truth is nimlangserver crashes often and I have never experienced nimsuggest crashing.

I kept the "official" nimlang plugin and used ```"nim.enableNimsuggest": true,```. Don't care that it's deprecated. It works. Better keep it around until nimlangserver works as well.

**Comment by jmgomez** [comment]
Can you post the steps to repro it?

**Comment by lewisl** [comment]
Um.  Boot vscode into a project folder, or the src folder of that project, and it fails. Not all the time. Hard to reprogram because it’s intermittent.

- Lewis

On Sep 5, 2024, at 4:32 AM, Juan M Gómez ***@***.***> wrote:

﻿

Can you post the steps to repro it?

**Comment by jmgomez** [comment]
Can you retest with the latest nimlangserver in head?

**Comment by lewisl** [comment]
How do I install the nimlangserver at the head commit on my machine?

Also, araq in another repo (the vscode plugin) suggested it could be related to config problems.  I readily agree I may have "things" (not sure which things...) configured wrong.  What config settings would be helpful to diagnose this?

**Comment by lewisl** [comment]
Ignore my previous reply.


  1.  I assume you mean only the nimlangserver, not nimlsp nor a different version of either of the nimlang VS code plugins.
  2.  If so, I RTFM which said: nimble install nimlangserver@#head

I’ll go back to the “supported” nim lang plugin, disable using nimsuggest directly and see what happens.

From: Juan M Gómez ***@***.***>
Reply-To: nim-lang/langserver ***@***.***>
Date: Tuesday, September 10, 2024 at 5:57 AM
To: nim-lang/langserver ***@***.***>
Cc: Lewis Levin ***@***.***>, Author ***@***.***>
Subject: Re: [nim-lang/langserver] connection to server is erroring (Issue #228)


Can you retest with the latest nimlangserver in head?

**Comment by lewisl** [comment]
F*ck nimble.

I did nimble install nimlangserver@#head

I got an entire new nim.  It messed up my paths.  It built the wrong architecture of nim. So, now I get to waste time to fix all that.

I try to work with you and do what you suggest, but no more.  Nim is great; the packages not so much.  Bad ecosystem is crippling the language.

Let me know when you have the issue fixed and you have done the testing, not me.

From: Juan M Gómez ***@***.***>
Reply-To: nim-lang/langserver ***@***.***>
Date: Tuesday, September 10, 2024 at 5:57 AM
To: nim-lang/langserver ***@***.***>
Cc: Lewis Levin ***@***.***>, Author ***@***.***>
Subject: Re: [nim-lang/langserver] connection to server is erroring (Issue #228)


Can you retest with the latest nimlangserver in head?

**Comment by jmgomez** [comment]
Relax. You came here complaining without a proper issue report. People is willing to help, but you should really take it more easy. Closing this as its a dead end.

**Comment by lewisl** [comment]
I provided what you asked for if I could.

I asked for what you needed when I knew I didn’t understand what you requested.

Nimble seems to wipe things out when installing other than a released version of a package.  Installing at head:

  *   Wiped out my homebrew installed nim.  Didn’t just change the paths to a different nim install—actually deleted the directories. No tool should ever delete directories for which it has no responsibility.
  *   Rebuilt nim using choosenim, which cannot build to arch: arm64.  This has been confirmed repeatedly but never fixed. Araq has expressed his skepticism about the tool, but lets its stand because he badly needs more helpers.
  *   As reported, all I did was ```nimble install nimlangserver@#head```. I assure I did not uninstall my working version of nim.

So I have rebuilt nim from source per the instructions on the website (about the 5th time in the past 2 weeks).  This works and creates a nim built with arch: arm64 that will compile executables for same.

Using the nimble built for arm64 by koch (per instructions), I installed the released nimlangserver. Now I am back to everything working again.

People who support nim often lament that nim doesn’t get the adoption it deserves (thanks mostly to Araq’s brilliance and nearly unhuman productivity).  You, on the other hand, have provided an object lesson why: the ecosystem lets the language down. Your stuff has some issues: it’s ok; software is hard and takes time.  Even araq implied he realized that maybe the lsp calling nimsuggest might be introducing some problems.

You can close it but the problem remains. I am not the only person reporting it.


From: Juan M Gómez ***@***.***>
Reply-To: nim-lang/langserver ***@***.***>
Date: Wednesday, September 11, 2024 at 2:18 PM
To: nim-lang/langserver ***@***.***>
Cc: Lewis Levin ***@***.***>, Author ***@***.***>
Subject: Re: [nim-lang/langserver] connection to server is erroring (Issue #228)


Closed #228<https://github.com/nim-lang/langserver/issues/228> as completed.
